streams/src/main/java/org/apache/kafka/streams/state/internals/ContextualRecord.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+
+import java.util.Objects;
+
+public class ContextualRecord<V> {
+    private final V value;
+    private final ProcessorRecordContext recordContext;
+
+    public ContextualRecord(final V value, final ProcessorRecordContext recordContext) {
+        this.value = value;
+        this.recordContext = recordContext;
+    }
+
+    public ProcessorRecordContext recordContext() {
+        return recordContext;
+    }
+
+    public V value() {
+        return value;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        final ContextualRecord<?> that = (ContextualRecord<?>) o;
+        return Objects.equals(value, that.value) &&
+            Objects.equals(recordContext, that.recordContext);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(value, recordContext);
+    }
+
+    @Override
+    public String toString() {
+        return "ContextualRecord{value=" + value + '}';
```

Comment:
`value` is always `byte[]` -- can we get a handle on the deserializer to get human readable output here? (one more reason to avoid generic if not necessary -- those issues slip easily with missing type information).

Reply:
I don't think the record needs to know how to deserialize itself.

Since `toString` is only for debugging, I'm fine printing out the `Arrays.toString` summary of the value.

If we wanted to print out the value in a log message, we would format it more specifically (including a deserialization if desired).

That said, I will go ahead and get rid of the generic type.

Reply:
> only for debugging

My point is, that even for debugging it's not useful to print `byte[]` -- my argument is, to either "fix this" or don't overwrite `toString()` at all.